posts/processing-sequential-javascript-promises-using-functional-composition/2020-03-26-processing-sequential-javascript-promises-using-functional-composition.md

```diff
@@ -0,0 +1,461 @@
+
+Let’s visualize how the piped function `getFollowersForUser(1)` will be executed.
+
+```jsx
+lookupUsers(lookupFollowerIDs(lookupUser(1)))
+```
+
+Hold-on a moment! Doesn’t it look different from the implementation of the nested promise? Yes, it is!
+
+> ### Instead of passing resolved value to the subsequent promise in the series, the actual promise object is passed.
+
+Because `lookupUser(1)` does not return the resolved value. It actually returns the Promise object itself.
+
+![Promise.resolve() returns the Promise object itself. Not the resolved value!](./images/promise-resolve-returns-promise.png)_Promise.resolve() returns the Promise object itself. Not the resolved value!_
+
+As a result, `lookupFollowerIDs` will receive a Promise object rather than `db.user` object.
+
+As per the definition, `lookupFollowerIDs` will run `promiseObject.followerIDs` rather than `user.followerIDs` which will return `undefined`
+
+Consequently, the last function in the series `lookUpUsers: (userIds = [])` will receive `undefined` rather than an array of `userIds` . This logical conclusion fits with the error we got when promises are composed using `pipe()`
+
+```jsx
+Uncaught TypeError: userIds.map is not a function
+```
+
+**Solution — **Programmatically chain the promises using `thenable` and `reduce()`
+
+## Sequential Promises As Closure + Functional Composition
+
+Consider the chained snippet we’ve created earlier.
+
+```jsx
+lookupUser(1)
+  .then(lookupFollowerIDs)
+  .then(lookupUsers)
+  .then(userFollowers => console.log(userFollowers))
+```
+
+This can also be re-written as,
+
+```jsx
+const promise1 = lookupUser(1).then(lookupFollowerIDs)
+
+const promise2 = promise1.then(lookupUsers)
+
+promise2.then(userFollowers => console.log(userFollowers))
+```
+
+As you may have noticed, at any given statement, a Promise object is chained with another Promise using `then()` .
+
+To implement this logic programmatically, as an initial step, a function can be created which receives two Promises and chain them using `then()`
+
+```jsx
+const internalPromisePipe = (f, g) =>
+  function() {
+    var ctx = this
+    return f.call(ctx, ...arguments).then(function(resolvedV) {
+      return g.call(ctx, resolvedV)
+    })
+  }
+```
+
+#### Why using call method to invoke Promise-returning functions?
+
+Preserving reference to `this` — Promise fulfilling function does not have access to `this` context as it creates its own context. Trying to use `this` without binding the context will result in *Uncaught Reference Error. *The `call` function invokes the argument function with provided `this` value (`ctx` in our case)
+
+### Applying promiseChainer() in pipe()
+
+```jsx
+const promisePipe = (...functions) => (...initialArgs) => {
+      return functions
+        .reduce((currFunc, nextFunc) => {
+          return **promiseChainer(currFunc, nextFunc);**
```

Comment:
@kalidasm Is the ** prefix and suffix intended?
I dont think so.

Reply:
Not intended.

Checked other posts for similar issues and fixed in #70